File: src/main/scala/org/dama/datasynth/runtime/spark/operators/matching/utils/traversals/RandomTraversal.scala
package org.dama.datasynth.runtime.spark.operators.matching.utils.traversals
import org.apache.spark.sql.DataFrame
import org.dama.datasynth.executionplan.ExecutionPlan
import org.dama.datasynth.executionplan.ExecutionPlan.EdgeTable

import scala.collection._
import scala.collection.mutable.ListBuffer
/**
  * Created by joangui on 18/07/2017.
  */
class RandomTraversal(edgeTable:DataFrame) extends Traversal{
  val targetNodes: Array[Long] = edgeTable.select(edgeTable("target")).collect.map(_.getAs[Long]("target"))
  val sourceNodes: Array[Long] = edgeTable.select(edgeTable("source")).collect.map(_.getAs[Long]("source"))
  val nodesTmp: Array[Long] = targetNodes.union(sourceNodes).distinct
  val nodesTmp2: mutable.ListBuffer[Long] = nodesTmp.foldLeft(new mutable.ListBuffer[Long])({case (nodes,nextNode)=>nodes += nextNode})
  val nodes: mutable.ListBuffer[Long] = util.Random.shuffle(nodesTmp2)
  override def hasNext(): Boolean = nodes.nonEmpty
  override def next(): Long = nodes.remove(0)
  def getNodes():ListBuffer[Long] = nodes

  //override def foreach[U](f: (Long) => U): Unit = nodes.foreach()
  override def foreach[U](f: Long => U) = nodes.foreach(f)

}

File: src/main/java/org/dama/datasynth/runtime/spark/operators/matching/graphs/RandomTraversal.java
package org.dama.datasynth.runtime.spark.operators.matching.graphs;

import org.dama.datasynth.runtime.spark.operators.matching.graphs.types.Graph;
import org.dama.datasynth.runtime.spark.operators.matching.graphs.types.Traversal;

import java.util.Collections;
import java.util.LinkedList;
import java.util.Random;

/**
 * Created by aprat on 12/03/17.
 */
public class RandomTraversal implements Traversal {

    private LinkedList<Long> nodes = null;
    @Override
    public void initialize(Graph graph) {
        nodes = new LinkedList<>(graph.getNodes());
        Random random = new Random(0L);
        Collections.shuffle(nodes,random);
    }

    @Override
    public boolean hasNext() {
        return !nodes.isEmpty();
    }

    @Override
    public Long next() {
        return nodes.pollFirst();
    }
}
